### PR TITLE
Fixes #13126

### LINKED ISSUE #13126: Github Discovery not filtering properly and eating our Github App API rate limit
Followed the docs here https://backstage.io/docs/integrations/github/discovery for setting up Github Discovery. I found it strange that I had to run `yarn add --cwd packages/backend @backstage/plugin-catalog-backend-module-github@next` instead of `yarn add --cwd packages/backend @backstage/plugin-catalog-backend-module-github` because the stable package didn't have `GitHubEntityProvider` in `@backstage/plugin-catalog-backend-module-github` but I moved forward anyway. 

Configured everything properly with following:

```
 providers:
    github:
      backstageCatalog:
        organization: 'REMOVED' # string
        catalogPath: '/groups/**/*.yaml' # this will search all folders for files that end in .yaml
        filters: # optional filters
          branch: 'develop' # optional string
          repository: 'backstage-catalog' # optional Regex
```

## Expected Behavior

I would expect this to look in one place only `https://github.com/REMOVED/backstage-catalog/tree/develop/groups/**/*.yaml`

## Actual Behavior

However, the logs show the following message:
```
[1] 2022-08-12T13:22:58.882Z catalog info Read 4276 GitHub repositories (173 matching the pattern) type=plugin target=github-provider:backstageCatalog class=GitHubEntityProvider taskId=github-provider:backstageCatalog:refresh taskInstanceId=8063561d-b4e4-4610-ac89-b630b7cd3feb
```

The logs then proceed to show the plugin scanning many repos in our org. Then we end up hitting our rate limit, which for a github app is over 15000 per hour.

## Steps to Reproduce

Follow the docs here (https://backstage.io/docs/integrations/github/discovery) to reproduce. This happens every time I start backstage. It's as if the filter pattern does not apply before scanning all repos.

## Context

This makes the suggested Github Discovery plugin setup unusable. I'm thinking of moving to the old Github Discovery setup that is flagged as (will be deprecated) because it worked properly.

- Output of `yarn backstage-cli info`: 

```text
➜ yarn backstage-cli info
yarn run v1.22.18
$ /Users/tim.harris/code/cfa/toolspod-backstage/node_modules/.bin/backstage-cli info
OS:   Darwin 21.6.0 - darwin/x64
node: v16.13.0
yarn: 1.22.18
cli:  0.17.2 (installed)

Dependencies:
  @backstage/app-defaults                          1.0.4
  @backstage/backend-common                        0.14.1, 0.15.0-next.2
  @backstage/backend-plugin-api                    0.1.0, 0.1.1-next.0
  @backstage/backend-tasks                         0.3.3, 0.3.4-next.0
  @backstage/catalog-client                        1.0.4
  @backstage/catalog-model                         1.1.0
  @backstage/cli-common                            0.1.9
  @backstage/cli                                   0.17.2
  @backstage/config-loader                         1.1.3
  @backstage/config                                1.0.1
  @backstage/core-app-api                          1.0.4
  @backstage/core-components                       0.10.0, 0.9.5
  @backstage/core-plugin-api                       1.0.4
  @backstage/errors                                1.1.0
  @backstage/integration-react                     1.1.2
  @backstage/integration                           1.2.2, 1.3.0-next.1
  @backstage/plugin-api-docs                       0.8.7
  @backstage/plugin-app-backend                    0.3.34
  @backstage/plugin-auth-backend                   0.14.1
  @backstage/plugin-auth-node                      0.2.3, 0.2.4-next.0
  @backstage/plugin-catalog-backend-module-github  0.1.6-next.2
  @backstage/plugin-catalog-backend-module-ldap    0.5.1
  @backstage/plugin-catalog-backend                1.3.0, 1.3.1-next.2
  @backstage/plugin-catalog-common                 1.0.4, 1.0.5-next.0
  @backstage/plugin-catalog-graph                  0.2.19
  @backstage/plugin-catalog-import                 0.8.10
  @backstage/plugin-catalog-node                   1.0.0, 1.0.1-next.0
  @backstage/plugin-catalog-react                  1.1.2
  @backstage/plugin-catalog                        1.4.0
  @backstage/plugin-github-actions                 0.5.7
  @backstage/plugin-org                            0.5.7
  @backstage/plugin-permission-common              0.6.3
  @backstage/plugin-permission-node                0.6.3, 0.6.4-next.0
  @backstage/plugin-permission-react               0.4.3
  @backstage/plugin-proxy-backend                  0.2.28
  @backstage/plugin-scaffolder-backend             1.4.0
  @backstage/plugin-scaffolder-common              1.1.2
  @backstage/plugin-scaffolder                     1.4.0
  @backstage/plugin-search-backend-module-pg       0.3.5
  @backstage/plugin-search-backend-node            0.6.2, 1.0.0
  @backstage/plugin-search-backend                 0.5.3
  @backstage/plugin-search-common                  0.3.5, 1.0.0
  @backstage/plugin-search-react                   0.2.1, 1.0.0
  @backstage/plugin-search                         0.9.0
  @backstage/plugin-tech-radar                     0.5.14
  @backstage/plugin-techdocs-backend               1.2.0
  @backstage/plugin-techdocs-module-addons-contrib 1.0.2
  @backstage/plugin-techdocs-node                  1.2.0
  @backstage/plugin-techdocs-react                 1.0.2
  @backstage/plugin-techdocs                       1.3.0
  @backstage/plugin-user-settings                  0.4.6
  @backstage/release-manifests                     0.0.4
  @backstage/test-utils                            1.1.2
  @backstage/theme                                 0.2.16
  @backstage/types                                 1.0.0
  @backstage/version-bridge                        1.0.1
✨  Done in 1.24s.
```

### PR DESCRIPTION
**Pull request by timharris777**
This PR is to fix this issue: https://github.com/backstage/backstage/issues/13126

### CODE REVIEW

**Review by awanlin — APPROVED** [review]
LGTM! Thanks for making this fix

File: plugins/catalog-backend-module-github/src/providers/GitHubEntityProvider.ts [hunk 1]
```diff
@@ -185,9 +185,12 @@ export class GitHubEntityProvider implements EntityProvider {
...
 
     const matchingRepositories = repositories.filter(r => {
-      return !r.isArchived && repositoryFilter
-        ? repositoryFilter?.test(r.name)
-        : true && r.defaultBranchRef?.name;
+      return (
+        !r.isArchived &&
+        (repositoryFilter
+          ? repositoryFilter?.test(r.name)
+          : true && r.defaultBranchRef?.name)
```
Comment (freben) [inline plugins/catalog-backend-module-github/src/providers/GitHubEntityProvider.ts:192]: Hm. Even after the fix, it's surprising to me that the old code checked for the presence of a main branch _only if_ there was no repository filter.

Comment (freben) [reply plugins/catalog-backend-module-github/src/providers/GitHubEntityProvider.ts:192]: I think this should be

```ts
!r.isArchived &&
(!repositoryFilter || repositoryFilter.test(r.name)) &&
r.defaultBranchRef?.name
```

**Review by freben — COMMENTED** [review]
Thanks! Just one final adjustment

**Review by freben — APPROVED** [review]

**Pull request merged**